ulib/experimental/FStar.OrdSet.fsti

```diff
@@ -35,8 +35,40 @@ val hasEq_ordset: a:eqtype -> f:cmp a
     [SMTPat (hasEq (ordset a f))]
 
 val empty        : #a:eqtype -> #f:cmp a -> Tot (ordset a f)
+
+val tail (#a:eqtype) (#f:cmp a) (s:ordset a f{s<>empty}) : ordset a f
+val head (#a:eqtype) (#f:cmp a) (s:ordset a f{s<>empty}) : a 
+
 val mem          : #a:eqtype -> #f:cmp a -> a -> s:ordset a f -> Tot bool
 
+(* currying-friendly version of mem, ready to be used as a lambda *)
+unfold let mem_of #a #f (s:ordset a f) x = mem x s
+
+val last (#a:eqtype) (#f:cmp a) (s: ordset a f{s <> empty}) 
+  : Tot (x:a{(forall (z:a{mem z s}). f z x) /\ mem x s})
+
+(* 
+  liat is the reverse of tail, i.e. a list of all but the last element.
+  A shortcut to (fst (unsnoc s)), which as a word is composed
+  in a remarkably similar fashion.
+*)
+val liat (#a:eqtype) (#f:cmp a) (s: ordset a f{s <> empty}) : Tot (l:ordset a f{
+    (forall x. mem x l = (mem x s && (x <> last s))) /\
+    (if tail s <> empty then (l <> empty) && (head s = head l) else true)
+  })
+
+val unsnoc (#a:eqtype) (#f:cmp a) (s: ordset a f{s <> empty}) : Tot (p:(ordset a f * a){
```

Comment:
Can it just be defined here itself as the pair? Then, the refinement is no longer necessary.

Reply:
yes, if the refinement is an equality, then let's just define it in the .fsti